src/PostgREST/App.hs

```diff
@@ -9,430 +9,664 @@ Some of its functionality includes:
+
+
+data RequestContext =
+  RequestContext
+    { rConfig      :: AppConfig
+    , rDbStructure :: DbStructure
+    , rApiRequest  :: ApiRequest
+    , rContentType :: ContentType
+    }
+
+type Handler = ExceptT Error
+
+type DbHandler = Handler SQL.Transaction
+
+-- | PostgREST application
+postgrest
+  :: Types.LogLevel
+  -> IORef AppConfig
+  -> IORef (Maybe DbStructure)
+  -> SQL.Pool
+  -> IO UTCTime
+  -> IO () -- ^ Lauch connection worker in a separate thread
+  -> Wai.Application
 postgrest logLev refConf refDbStructure pool getTime connWorker =
-  pgrstMiddleware logLev $ \ req respond -> do
-    time <- getTime
-    body <- strictRequestBody req
-    maybeDbStructure <- readIORef refDbStructure
-    conf <- readIORef refConf
-    case maybeDbStructure of
-      Nothing -> respond . errorResponseFor $ ConnectionLostError
-      Just dbStructure -> do
-        response <- do
-          let apiReq = userApiRequest (configDbSchemas conf) (configDbRootSpec conf) dbStructure req body
-          case apiReq of
-            Left err -> return . errorResponseFor $ err
-            Right apiRequest -> do
-              -- The jwt must be checked before touching the db.
-              attempt <- attemptJwtClaims (configJWKS conf) (configJwtAudience conf) (toS $ iJWT apiRequest) time (configJwtRoleClaimKey conf)
-              case jwtClaims attempt of
-                Left errJwt -> return . errorResponseFor $ errJwt
-                Right claims -> do
-                  let
-                    authed = containsRole claims
-                    shouldCommit   = configDbTxAllowOverride conf && iPreferTransaction apiRequest == Just Commit
-                    shouldRollback = configDbTxAllowOverride conf && iPreferTransaction apiRequest == Just Rollback
-                    preferenceApplied
-                      | shouldCommit    = addHeadersIfNotIncluded [(hPreferenceApplied, BS.pack (show Commit))]
-                      | shouldRollback  = addHeadersIfNotIncluded [(hPreferenceApplied, BS.pack (show Rollback))]
-                      | otherwise       = identity
-                    handleReq = do
-                      when (shouldRollback || (configDbTxRollbackAll conf && not shouldCommit)) HT.condemn
-                      mapResponseHeaders preferenceApplied <$> runPgLocals conf claims (app dbStructure conf) apiRequest
-                  dbResp <- P.use pool $ HT.transaction HT.ReadCommitted (txMode apiRequest) handleReq
-                  return $ either (errorResponseFor . PgError authed) identity dbResp
-        -- Launch the connWorker when the connection is down. The postgrest function can respond successfully(with a stale schema cache) before the connWorker is done.
-        when (responseStatus response == status503) connWorker
+  Middleware.pgrstMiddleware logLev $
+    \req respond ->
+      do
+        time <- getTime
+        conf <- readIORef refConf
+        maybeDbStructure <- readIORef refDbStructure
```

Comment:
Is it possible to handle the `throwError Error.ConnectionLostError` here already and return `dbStructure` instead of `maybeDbStructure`?

Reply:
Yes, but it would be difficult to do without changing the behavior of when the connectionWorker is launched (https://github.com/monacoremo/postgrest/blob/a6ba8784e9096eee2e30f3c049ec7eeec9886a7c/src/PostgREST/App.hs#L90), which I'm not sure is fully correct right now, but I wouldn't want to change it in this PR.